compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaLexer.java

```diff
@@ -543,9 +555,133 @@ private STToken processNumericLiteral(int startChar) {
 
     /**
      * <p>
-     * Process and returns a hex integer literal.
+     * Process and returns a decimal floating point literal.
      * </p>
      * <code>
+     * DecimalFloatingPointNumber :=
+     *    DecimalNumber Exponent [FloatingPointTypeSuffix]
+     *    | DottedDecimalNumber [Exponent] [FloatingPointTypeSuffix]
+     *    | DecimalNumber FloatingPointTypeSuffix
+     * <br/>
+     * DottedDecimalNumber := DecimalNumber . Digit* | . Digit+
+     * <br/>
+     * FloatingPointTypeSuffix := DecimalTypeSuffix | FloatTypeSuffix
+     * <br/>
+     * DecimalTypeSuffix := d | D
+     * <br/>
+     * FloatTypeSuffix :=  f | F
+     * </code>
+     *
+     * @return The decimal floating point literal.
+     */
+    private STToken processDecimalFloatLiteral() {
+        boolean exponent = false;
+        boolean dot = false;
+        int nextChar;
+        while (true) {
+            nextChar = peek();
+            switch (nextChar) {
+                case 'f':
+                case 'F':
+                case 'd':
+                case 'D':
+                    reader.advance();
+                    break;
+                case 'e':
+                case 'E':
+                    if (!exponent) { // Check whether an exponent has been passed previously
+                        if (processExponent()) {
```

Comment:
Can we avoid these if-conditions, and the flags?

Looking at the grammar, there are three main components the decimalNumber,  the exponent, and the suffix. These always exists in order. For example: decimalNumber component always the first component. suffix always comes after the decimalNumber, and always before the suffix.

So, what I feel is, suppose we have three separate functions to process these three separate components, (say `a()`, `b()` and `c()` ) if we chain these three, then the complexity would be reduced.

for e.g: we start from `a()` and continue till we reach the exponent, and once we reach the exponent, call `b()`. But never return to a() and process again. Simply finish the processing and exit from b(). So `a()` never have to worry whether `b()` was successfully completed or not..

Reply:
I'll make that change.